### ISSUE TITLE
bad light userdata pointer LuaJIT 64 bits

### ISSUE THREAD

**Issue by sonoro1234** [LuaLanes/lanes#147]
Using lanes from a 64 bits compilation of LuaJIT gives the error`bad light userdata pointer`

I have a plan to solve it by replacing push_lightuserdata with another function when compiling with LuaJIT 64 bits
see: LuaJIT/LuaJIT#435

in lua-nginx they did:
https://github.com/openresty/lua-nginx-module/issues/1152#issuecomment-384069881

There are several solutions depending on the use case of lua_pushlightuserdata:

where is it is used for pushing  a unique key (DEEP_LOOKUP_KEY) it could be used a predefined string as the key substitution
but there are other cases. Isn't it?

For example
https://github.com/LuaLanes/lanes/blob/master/src/keeper.c#L196
would need wrapping in a userdata or with luaL_ref

What is your general impression?

**Comment by benoit-germain** [comment]
So basically what happens is that when compiling for luaJIT, lua_pushlightuserdata(L, p) will only store 47 bits of the pointer, causing lua_touserdata() to compare different to p?

**Comment by sonoro1234** [comment]
Also, LuaJIT exits with error `bad light userdata pointer` when it does not fit in 47 bits

**Comment by sonoro1234** [comment]
Here is a method of wrapping in a userdata, althought is not for saving as table value could be an inspiration
https://github.com/ahupowerdns/luawrapper/pull/45/commits/70dd55252bc9534544166d0cddf0f06445cae4aa

**Comment by sonoro1234** [comment]
Perhaps a branch should be created and give a try to lua_pushlstring(L,pointer,sizeof(pointer)) as a general replacement. It would be slower but can be used as a transition test.

**Comment by benoit-germain** [comment]
As a matter of fact, Universe is a full userdata keyed by some light userdata constant, so just making that constant only having 47 significant bits would solve the LuaJIT compatibility issue. With any luck the rest can be taken care of in a similar fashion. I'll investigate some more.

**Comment by sonoro1234** [comment]
with this substitution I am getting a crash in 32 and 64 bits
```
void pushLUD(lua_State *L, void *p)
{
	lua_pushlstring(L,p,sizeof(p));
	//lua_pushlightuserdata(L,p);
}
void* touserdata(lua_State *L,int index)
{
	size_t len;
	const char * ptrs = lua_tolstring(L,index,&len);
	assert(len == sizeof(void *));
	return ptrs;
	//return lua_touserdata(L,index);
}
```

**Comment by benoit-germain** [comment]
This will get you a different result on each Lane since each Lua state will intern the "string" in a different buffer.
Maybe it will work better with (untested)
```
void* touserdata(lua_State *L,int index)
{
	size_t len;
	const char * ptrs = lua_tolstring(L,index,&len);
	assert(len == sizeof(void *));
	return *(void**)ptrs;
	//return lua_touserdata(L,index);
}
```

**Comment by sonoro1234** [comment]
the crash was in lua_pushlstring 
I think it should be
```
void pushLUD(lua_State *L, void *p)
{
	lua_pushlstring(L,&p,sizeof(p));
	//lua_pushlightuserdata(L,p);
}
```
But then I get

```
Program received signal SIGSEGV, Segmentation fault.
0x618c4170 in LG_configure (L=0x3201c0) at C:\luaGL\gitsources\lanes\src\lanes.c:3134
3134                    U->timer_deep = *(DeepPrelude**) touserdata( L, -1);
(gdb) thread apply all bt

Thread 1 (Thread 5956.0x14c4):
#0  0x618c4170 in LG_configure (L=0x3201c0) at C:\luaGL\gitsources\lanes\src\lanes.c:3134
#1  0x66d82d5a in lj_BC_FUNCC () from C:\luaGL\gitsources\buildluajit2.1debug\lua51.dll
#2  0x003201c0 in ?? ()
#3  0x00010662 in ?? ()
#4  0x00000c02 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
(gdb)
```

**Comment by sonoro1234** [comment]
Now with
```
void* touserdata(lua_State *L,int index)
{
	size_t len;
	const char * ptrs = lua_tolstring(L,index,&len);
	printf("len %d sizevoid %d ptrs %p\n",len,sizeof(void*),ptrs);
	if (!ptrs)
		return ptrs;
	void * ptr;
	memcpy(&ptr,ptrs,len);
	return ptr;
}
```
says:
```
Starting program: C:\luaGL\gitsources\buildluajit2.1debug\luajit.exe C:\luaGL\gitsources\lanes\memory_leak.lua

[New Thread 4144.0x1fc]
len 0 sizevoid 4 ptrs 00000000
len 0 sizevoid 4 ptrs 00000000
len 0 sizevoid 4 ptrs 00000000
len 0 sizevoid 4 ptrs 00000000

Program received signal SIGSEGV, Segmentation fault.
0x618c4170 in LG_configure (L=0x3a01c0) at C:\luaGL\gitsources\lanes\src\lanes.c:3134
3134                    U->timer_deep = *(DeepPrelude**) touserdata( L, -1);
(gdb) thread apply all bt

Thread 1 (Thread 4144.0x1fc):
#0  0x618c4170 in LG_configure (L=0x3a01c0) at C:\luaGL\gitsources\lanes\src\lanes.c:3134
#1  0x66d82d5a in lj_BC_FUNCC () from C:\luaGL\gitsources\buildluajit2.1debug\lua51.dll
#2  0x003a01c0 in ?? ()
#3  0x00010662 in ?? ()
#4  0x00000c02 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
(gdb)

```
Which seems caused by defererencing a NULL pointer

**Comment by sonoro1234** [comment]
detection only fails in universe.c
```
Scanning dependencies of target core
[ 12%] Building C object CMakeFiles/core.dir/src/lanes.c.obj
In file included from C:\luaGL\gitsources\lanes\src\tools.h:8,
                 from C:\luaGL\gitsources\lanes\src\lanes.c:93:
C:\luaGL\gitsources\lanes\src\macros_and_utils.h:16:9: note: #pragma message: LuaJIT 64 bits detected: don't p
ropagate allocf
 #pragma message( "LuaJIT 64 bits detected: don't propagate allocf")
         ^~~~~~~
[ 25%] Building C object CMakeFiles/core.dir/src/deep.c.obj
In file included from C:\luaGL\gitsources\lanes\src\tools.h:8,
                 from C:\luaGL\gitsources\lanes\src\deep.c:45:
C:\luaGL\gitsources\lanes\src\macros_and_utils.h:16:9: note: #pragma message: LuaJIT 64 bits detected: don't p
ropagate allocf
 #pragma message( "LuaJIT 64 bits detected: don't propagate allocf")
         ^~~~~~~
[ 37%] Building C object CMakeFiles/core.dir/src/threading.c.obj
[ 50%] Building C object CMakeFiles/core.dir/src/tools.c.obj
In file included from C:\luaGL\gitsources\lanes\src\tools.h:8,
                 from C:\luaGL\gitsources\lanes\src\tools.c:42:
C:\luaGL\gitsources\lanes\src\macros_and_utils.h:19:9: note: #pragma message: PUC-Lua detected: propagate allo
cf
 #pragma message( "PUC-Lua detected: propagate allocf")
         ^~~~~~~
C:\luaGL\gitsources\lanes\src\tools.c: In function 'touserdata':
C:\luaGL\gitsources\lanes\src\tools.c:63:10: warning: return discards 'const' qualifier from pointer target ty
pe [-Wdiscarded-qualifiers]
   return ptrs;
          ^~~~
[ 62%] Building C object CMakeFiles/core.dir/src/keeper.c.obj
In file included from C:\luaGL\gitsources\lanes\src\tools.h:8,
                 from C:\luaGL\gitsources\lanes\src\keeper.c:49:
C:\luaGL\gitsources\lanes\src\macros_and_utils.h:16:9: note: #pragma message: LuaJIT 64 bits detected: don't p
ropagate allocf
 #pragma message( "LuaJIT 64 bits detected: don't propagate allocf")
         ^~~~~~~
[ 75%] Building C object CMakeFiles/core.dir/src/compat.c.obj
[ 87%] Building C object CMakeFiles/core.dir/src/universe.c.obj
In file included from C:\luaGL\gitsources\lanes\src\tools.h:8,
                 from C:\luaGL\gitsources\lanes\src\universe.h:9,
                 from C:\luaGL\gitsources\lanes\src\universe.c:31:
C:\luaGL\gitsources\lanes\src\macros_and_utils.h:19:9: note: #pragma message: PUC-Lua detected: propagate allo
cf
 #pragma message( "PUC-Lua detected: propagate allocf")
         ^~~~~~~
[100%] Linking C shared module core.dll
[100%] Built target core
```

**Comment by sonoro1234** [comment]
Runs in LuaJIT 64 and posts
```
C:\luaGL\gitsources\buildluajit2.164bits\luajit.exe: C:\luaGL\gitsources\lanes\memory_leak.lua:9: luaG_newstate() failed while creating state; out of memory
stack traceback:
        [C]: at 0x618c1780
        C:\luaGL\gitsources\lanes\memory_leak.lua:9: in main chunk
        [C]: at 0x00402070
```
with the script in issue #141 
but this could be the another issue with LuaJIT 64 memory

**Comment by sonoro1234** [comment]
Debugging test/basic.lua shows that sometimes len != sizeof(void *) but as I catch it in an assert I loose all the backtrace (How to stop the program on this without assert?)

**Comment by sonoro1234** [comment]
But with this modification (len!=8 returns null) no SIGVSEG and
```
        ---=== Communications ===---


[New Thread 183620.0x2cd64]
1 sent
Lane starts!
2 sent
.1 received
.[Thread 183620.0x2cd64 exited with code 0]
..................................................................................................3 sent
lane status: error
nil     C:\luaGL\gitsources\lanes\tests\basic.lua:215: assertion failed!        table: 0x002d8470
                                ---> lane 'chunk' collected with status closed
                                ---> lane 'task(1,999999999,1)' collected with status closed
C:\luaGL\gitsources\buildluajit2.164bits\luajit.exe: C:\luaGL\gitsources\lanes\tests\basic.lua:262: assertion
failed!
stack traceback:
        [C]: in function 'error'
        c:/luaGL/gitsources/lanes/tests/assert.lua:90: in function '_assert'
        c:/luaGL/gitsources/lanes/tests/assert.lua:122: in function 'assert'
        C:\luaGL\gitsources\lanes\tests\basic.lua:262: in main chunk
        [C]: at 0x00402070
[Inferior 1 (process 183620) exited with code 01]
```

**Comment by sonoro1234** [comment]
with last commit all the tests worked!!

Now there is only the problem of confusing a string of len 8 with a pretended lightuserdata.
Instead of just saving the pointer as a length 8 string perhaps would be better as a length 16 string with the first 8 chars being the c pointer and the last 8 ones being a "magic" string

**Comment by benoit-germain** [comment]
To be fair I don't know if trying to have an alternate way of managing light userdata is all that useful. After all, LuaJIT x64 users will already experience similar lightuserdata issues in other module without Lanes. And as far as Lanes is concerned, lightuserdata are just here to make lightweight table keys and "constants" such as cancel_error. The idea is that a pure Lua program can't generate such a value by itself, and so can't create a collision by accident. Also, Lane's use of those keys isn't likely to create collisions with other modules since these keys are used in the registry, and even for the most part in private Lanes tables. Of course, the higher the number of bits the less risk of collisions.
My point is: Lanes has to be adapted as you suggest by having function to get/set unique keys. But I feel that just still having a lightuserdata under the hood and simply fixing the code so that all constants have 47 significant bits is just enough and won't create any confusion with those special strings.

**Comment by sonoro1234** [comment]
> After all, LuaJIT x64 users will already experience similar lightuserdata issues in other module without Lanes

Yes, this is an very active issue in several modules: torch, nginx,LPEG,PowerDNS,openresty,...
I have not seen an unified solution for this so any experimentation is needed

> But I feel that just still having a lightuserdata under the hood and simply fixing the code so that all constants have 47 significant bits is just enough and won't create any confusion with those special strings.

Wouldnt it make pointers pushed as lightuserdata invalid?

**Comment by benoit-germain** [comment]
Of course. I don't know about other modules, but such pointers aren't dereferenced by Lanes to actually use the pointed objects. It's just a 64 bits value that a pure Lua program can't generate, that Lanes uses as unique identifiers. Not using constants but function pointers is just an easy means of having different values without hardcoding them, and ensuring unicity inside Lanes (and even possibly those values are different with each run, thus strengthening Lanes against attacks).

**Comment by sonoro1234** [comment]
Test seem to run ok with
```
void pushLUD(lua_State *L, void *p)
{
	void *p2 = (void *) ((uintptr_t) p & (uintptr_t) 0x7FFFFFFFFFFF);
	lua_pushlightuserdata(L,p2);
}
void* touserdata(lua_State *L,int index)
{
	return lua_touserdata(L,index);
}
```

**Comment by sonoro1234** [comment]
I have just checked that the only pointers greater than 47 bits are defined as a crc number:

DEEP_LOOKUP_KEY
DEEP_PROXY_CACHE_KEY and
UNIVERSE_REGKEY

So it seems that they are the only ones causing this issue and should be changed

We have not  the VA48bit issue (only happens in arm64) or the x86_64 VA57bits issue that is expected to appear in new machines

**Comment by benoit-germain** [comment]
Well to do things properly my idea would be to have a pair of special functions for "unique internal key"  that Lanes is generating/using, and having the code use them everywhere such a key is used.
Also, I also have places where I still use function pointers, such as REG_MTID. This will break if the pointer has more than 47 significant bits and should be changed to be masked properly (or to some hardcoded constant like the others).

**Comment by sonoro1234** [comment]
In the meanwhile I think it should be usefull to set constants

DEEP_LOOKUP_KEY
DEEP_PROXY_CACHE_KEY and
UNIVERSE_REGKEY

to any 47 bits value so that LuaJIT 64  can be used

**Comment by benoit-germain** [comment]
Well, I have stuff to commit/merge, but unfortunately for some reason my home PC's Github desktop doesn't show me the last commits I did with my work PC (latest shown is 95ca27b5658c9b4da4ec67f6e922ff370eae05e5). Until I can solve this I won't commit anything since I am out of office till the end of the week...

**Comment by sonoro1234** [comment]
Perhap you have to git pull?

**Comment by benoit-germain** [comment]
Uh oh. Ensuring 64 bits LuaJIT operation will mean I need to break deep userdata implementation API. See
```
void free_deep_prelude( lua_State* L, DeepPrelude* prelude_)
{
	// Call 'idfunc( "delete", deep_ptr )' to make deep cleanup
	lua_pushlightuserdata( L, prelude_->deep);
	ASSERT_L( prelude_->idfunc);
	prelude_->idfunc( L, eDO_delete);
	DEEP_FREE( (void*) prelude_);
}
```

This is because currently the deep userdata idfunc grabs the deep object's pointer through the Lua stack. Which means that the pointer must fit in the 47 bits. I have to change this so that the pointer is provided as a regular C function argument.

**Comment by sonoro1234** [comment]
It seems that when a pointer is from the heap (created with malloc) it will fit in 47 bits (until 64 bit systems with 57bit VA appear). So it is not still a problem (In my tests only fail with crc64 generated values)

**Comment by sonoro1234** [comment]
with last https://github.com/LuaLanes/lanes/commit/ea9e8a3af1c2357c454ef18c8136c14a22b8675a
still not working on LuaJIT 64 (crc64 generated values)

**Comment by benoit-germain** [comment]
Its normal, it was unrelated :-).
d82bf70e9db3a25ec451b599660f3a837f21caee should take care of it.

**Comment by sonoro1234** [comment]
Building last master 32bits ( I dont know 64bits)
```
C:\luaGL\gitsources\build_lanes>mingw32-make install
Scanning dependencies of target core
[ 12%] Building C object CMakeFiles/core.dir/src/lanes.c.obj
[ 25%] Building C object CMakeFiles/core.dir/src/deep.c.obj
[ 37%] Building C object CMakeFiles/core.dir/src/tools.c.obj
[ 50%] Building C object CMakeFiles/core.dir/src/keeper.c.obj
[ 62%] Building C object CMakeFiles/core.dir/src/universe.c.obj
C:\luaGL\gitsources\lanes\src\universe.c:37:63: error: macro "DECLARE_UNIQUE_KEY" passed 2 arguments, but take
s just 1
 static DECLARE_UNIQUE_KEY( UNIVERSE_REGKEY, 0x9f877b2cf078f17f);
                                                               ^
In file included from C:\luaGL\gitsources\lanes\src\universe.c:34:0:
C:\luaGL\gitsources\lanes\src\universe.c: In function 'universe_create':
C:\luaGL\gitsources\lanes\src\universe.c:45:22: error: 'UNIVERSE_REGKEY' undeclared (first use in this functio
n)
  push_unique_key( L, UNIVERSE_REGKEY);                                                          // universe U
NIVERSE_REGKEY
                      ^
C:\luaGL\gitsources\lanes\src\uniquekey.h:22:61: note: in definition of macro 'push_unique_key'
 #define push_unique_key( L, key_) lua_pushlightuserdata( L, key_.value)
                                                             ^
C:\luaGL\gitsources\lanes\src\universe.c:45:22: note: each undeclared identifier is reported only once for eac
h function it appears in
  push_unique_key( L, UNIVERSE_REGKEY);                                                          // universe U
NIVERSE_REGKEY
                      ^
C:\luaGL\gitsources\lanes\src\uniquekey.h:22:61: note: in definition of macro 'push_unique_key'
 #define push_unique_key( L, key_) lua_pushlightuserdata( L, key_.value)
                                                             ^
C:\luaGL\gitsources\lanes\src\universe.c: In function 'universe_store':
C:\luaGL\gitsources\lanes\src\universe.c:56:22: error: 'UNIVERSE_REGKEY' undeclared (first use in this functio
n)
  push_unique_key( L, UNIVERSE_REGKEY);
                      ^
C:\luaGL\gitsources\lanes\src\uniquekey.h:22:61: note: in definition of macro 'push_unique_key'
 #define push_unique_key( L, key_) lua_pushlightuserdata( L, key_.value)
                                                             ^
C:\luaGL\gitsources\lanes\src\universe.c: In function 'universe_get':
C:\luaGL\gitsources\lanes\src\universe.c:69:22: error: 'UNIVERSE_REGKEY' undeclared (first use in this functio
n)
  push_unique_key( L, UNIVERSE_REGKEY);
                      ^
C:\luaGL\gitsources\lanes\src\uniquekey.h:22:61: note: in definition of macro 'push_unique_key'
 #define push_unique_key( L, key_) lua_pushlightuserdata( L, key_.value)
                                                             ^
CMakeFiles\core.dir\build.make:212: recipe for target 'CMakeFiles/core.dir/src/universe.c.obj' failed
mingw32-make[2]: *** [CMakeFiles/core.dir/src/universe.c.obj] Error 1
CMakeFiles\Makefile2:66: recipe for target 'CMakeFiles/core.dir/all' failed
mingw32-make[1]: *** [CMakeFiles/core.dir/all] Error 2
Makefile:126: recipe for target 'all' failed
mingw32-make: *** [all] Error 2
```

**Comment by benoit-germain** [comment]
How can the M$ compiler be so broken and not showing that...

**Comment by sonoro1234** [comment]
Solved in 32 bits but now 64bits
```
C:\luaGL\gitsources\build_lanes64>mingw32-make install
Scanning dependencies of target core
[ 12%] Building C object CMakeFiles/core.dir/src/lanes.c.obj
In file included from C:\luaGL\gitsources\lanes\src\keeper.h:6,
                 from C:\luaGL\gitsources\lanes\src\lanes.c:95:
C:\luaGL\gitsources\lanes\src\uniquekey.h:14:43: error: invalid operands to binary & (have 'void *' and 'long
long unsigned int')
 #define MAKE_UNIQUE_KEY( p_) ((void*)(p_) & 0x7fffffffffffull)
                                           ^
C:\luaGL\gitsources\lanes\src\uniquekey.h:20:72: note: in expansion of macro 'MAKE_UNIQUE_KEY'
 #define DECLARE_CONST_UNIQUE_KEY( name_, p_) UniqueKey const name_ = { MAKE_UNIQUE_KEY( p_)}
                                                                        ^~~~~~~~~~~~~~~
C:\luaGL\gitsources\lanes\src\keeper.h:39:8: note: in expansion of macro 'DECLARE_CONST_UNIQUE_KEY'
 static DECLARE_CONST_UNIQUE_KEY( NIL_SENTINEL, 0x7eaafa003a1d11a1);
        ^~~~~~~~~~~~~~~~~~~~~~~~
C:\luaGL\gitsources\lanes\src\uniquekey.h:14:43: error: invalid operands to binary & (have 'void *' and 'long
long unsigned int')
 #define MAKE_UNIQUE_KEY( p_) ((void*)(p_) & 0x7fffffffffffull)
                                           ^
C:\luaGL\gitsources\lanes\src\uniquekey.h:20:72: note: in expansion of macro 'MAKE_UNIQUE_KEY'
 #define DECLARE_CONST_UNIQUE_KEY( name_, p_) UniqueKey const name_ = { MAKE_UNIQUE_KEY( p_)}
                                                                        ^~~~~~~~~~~~~~~
C:\luaGL\gitsources\lanes\src\lanes.c:198:8: note: in expansion of macro 'DECLARE_CONST_UNIQUE_KEY'
 static DECLARE_CONST_UNIQUE_KEY( CANCEL_TEST_KEY, 0xe66f5960c57d133a); // used as registry key
        ^~~~~~~~~~~~~~~~~~~~~~~~
C:\luaGL\gitsources\lanes\src\uniquekey.h:14:43: error: invalid operands to binary & (have 'void *' and 'long
long unsigned int')
 #define MAKE_UNIQUE_KEY( p_) ((void*)(p_) & 0x7fffffffffffull)
                                           ^
C:\luaGL\gitsources\lanes\src\uniquekey.h:20:72: note: in expansion of macro 'MAKE_UNIQUE_KEY'
 #define DECLARE_CONST_UNIQUE_KEY( name_, p_) UniqueKey const name_ = { MAKE_UNIQUE_KEY( p_)}
                                                                        ^~~~~~~~~~~~~~~
C:\luaGL\gitsources\lanes\src\lanes.c:245:8: note: in expansion of macro 'DECLARE_CONST_UNIQUE_KEY'
 static DECLARE_CONST_UNIQUE_KEY( CANCEL_ERROR, 0xe97d41626cc97577); // 'cancel_error' sentinel
        ^~~~~~~~~~~~~~~~~~~~~~~~
C:\luaGL\gitsources\lanes\src\uniquekey.h:14:43: error: invalid operands to binary & (have 'void *' and 'long
long unsigned int')
 #define MAKE_UNIQUE_KEY( p_) ((void*)(p_) & 0x7fffffffffffull)
                                           ^
C:\luaGL\gitsources\lanes\src\uniquekey.h:20:72: note: in expansion of macro 'MAKE_UNIQUE_KEY'
 #define DECLARE_CONST_UNIQUE_KEY( name_, p_) UniqueKey const name_ = { MAKE_UNIQUE_KEY( p_)}
                                                                        ^~~~~~~~~~~~~~~
C:\luaGL\gitsources\lanes\src\lanes.c:267:8: note: in expansion of macro 'DECLARE_CONST_UNIQUE_KEY'
 static DECLARE_CONST_UNIQUE_KEY( STACK_TRACE_KEY, 0x024d5411677ce879);
        ^~~~~~~~~~~~~~~~~~~~~~~~
C:\luaGL\gitsources\lanes\src\uniquekey.h:14:43: error: invalid operands to binary & (have 'void *' and 'long
long unsigned int')
 #define MAKE_UNIQUE_KEY( p_) ((void*)(p_) & 0x7fffffffffffull)
                                           ^
C:\luaGL\gitsources\lanes\src\uniquekey.h:20:72: note: in expansion of macro 'MAKE_UNIQUE_KEY'
 #define DECLARE_CONST_UNIQUE_KEY( name_, p_) UniqueKey const name_ = { MAKE_UNIQUE_KEY( p_)}
                                                                        ^~~~~~~~~~~~~~~
C:\luaGL\gitsources\lanes\src\lanes.c:1825:8: note: in expansion of macro 'DECLARE_CONST_UNIQUE_KEY'
 static DECLARE_CONST_UNIQUE_KEY( EXTENDED_STACK_TRACE_REGKEY, 0x7a59821071066e49); // used as registry key
        ^~~~~~~~~~~~~~~~~~~~~~~~
C:\luaGL\gitsources\lanes\src\lanes.c: In function 'LG_set_debug_threadname':
C:\luaGL\gitsources\lanes\src\uniquekey.h:14:43: error: invalid operands to binary & (have 'void *' and 'long
long unsigned int')
 #define MAKE_UNIQUE_KEY( p_) ((void*)(p_) & 0x7fffffffffffull)
                               ~~~~~~~~~~~ ^
C:\luaGL\gitsources\lanes\src\uniquekey.h:20:72: note: in expansion of macro 'MAKE_UNIQUE_KEY'
 #define DECLARE_CONST_UNIQUE_KEY( name_, p_) UniqueKey const name_ = { MAKE_UNIQUE_KEY( p_)}
                                                                        ^~~~~~~~~~~~~~~
C:\luaGL\gitsources\lanes\src\lanes.c:1970:2: note: in expansion of macro 'DECLARE_CONST_UNIQUE_KEY'
  DECLARE_CONST_UNIQUE_KEY( hidden_regkey, LG_set_debug_threadname);
  ^~~~~~~~~~~~~~~~~~~~~~~~
C:\luaGL\gitsources\lanes\src\lanes.c: At top level:
C:\luaGL\gitsources\lanes\src\uniquekey.h:14:43: error: invalid operands to binary & (have 'void *' and 'long
long unsigned int')
 #define MAKE_UNIQUE_KEY( p_) ((void*)(p_) & 0x7fffffffffffull)
                                           ^
C:\luaGL\gitsources\lanes\src\uniquekey.h:20:72: note: in expansion of macro 'MAKE_UNIQUE_KEY'
 #define DECLARE_CONST_UNIQUE_KEY( name_, p_) UniqueKey const name_ = { MAKE_UNIQUE_KEY( p_)}
                                                                        ^~~~~~~~~~~~~~~
C:\luaGL\gitsources\lanes\src\lanes.c:2195:8: note: in expansion of macro 'DECLARE_CONST_UNIQUE_KEY'
 static DECLARE_CONST_UNIQUE_KEY( GCCB_KEY, 0xcfb1f046ef074e88);
        ^~~~~~~~~~~~~~~~~~~~~~~~
mingw32-make[2]: *** [CMakeFiles\core.dir\build.make:63: CMakeFiles/core.dir/src/lanes.c.obj] Error 1
mingw32-make[1]: *** [CMakeFiles\Makefile2:67: CMakeFiles/core.dir/all] Error 2
mingw32-make: *** [Makefile:127: all] Error 2

```

**Comment by sonoro1234** [comment]
Thanks Benoit,
Works perfect now

Only left to do https://github.com/LuaLanes/lanes/issues/141#issuecomment-433811208

ffi.abi("64bit") true and ffi.abi"gc64" false -> protect_allocator=false
ffi.abi("64bit") true and ffi.abi"gc64" true -> protect_allocator=true
ffi.abi("32bit") true -> protect_allocator=true

for the protect_allocator issue(probably from lanes.lua)
